### ISSUE TITLE
RabbitMQ Component is not closing/disposing connections during Sidecar Component change/reload

### ISSUE THREAD

**Issue by slacksach** [dapr/components-contrib#1318]
## Expected Behavior
Dapr should dispose of RabbitMQ connections or re-use where applicable

## Actual Behavior
Over the course of 3 hours, we observed a 4x growth in active connections to RabbitMQ when using Dapr 1.4.3

This was able to cause a Production RabbitMQ outage on a 3-Node RabbitMQ Cluster

## Steps to Reproduce the Problem
1. Create a pod with a Dapr sidecar and 10x topics, each with a different pubsub
2. View the RabbitMQ connection count (either in the management UI or via Grafana dashboard, rabbitmqctl etc
3. Observe that every N minutes, the number of connections doubles

## Release Note
Fixed RabbitMQ connection close issue

**Comment by slacksach** [comment]
cc @yaron2 @pkedy @dmitsh - Since it is nearly 2am UK time, I'll add some more details in the morning.

**Comment by slacksach** [comment]
closing as this is fixed in - https://github.com/dapr/components-contrib/issues/1187

**Comment by slacksach** [comment]
@yaron2 @pkedy @dmitsh - This appears to still be an issue in Dapr v1.5.0, we upgraded from v1.4.3 and we're still seeing this exponential connection growth in Production, we've separated out the VHOST so we're now sure it is coming from the Dapr RabbitMQ component side.

Environment
Kubernetes 1.21.2
RabbitMQ 3.8.9 via Operator
Dapr v1.5.0

50 or so separate pubsub components (To work around the existing issue in Dapr RabbitMQ with rogue messages)

After 15 minutes,
![image](https://user-images.githubusercontent.com/75119677/142843469-20b1478c-5b9f-4ae0-bffd-1af0bd3de68c.png)

After restarting all our Dapr deployments
![image](https://user-images.githubusercontent.com/75119677/142843689-bcb7afce-14c1-4f75-8287-946543b9fc1a.png)

**Comment by slacksach** [comment]
> Can you point to the issue of "rogue messages"?

@yaron2 sure, https://github.com/dapr/components-contrib/issues/826, this one is still a major issue for us, the only workaround is to use 1 pubsub component per topic otherwise we cannot run multiple replicas of pods.
Unfortunately running 1 pubsub per topic now highlights the connection closing/disposing issue, at the moment we're just having to keep an eye on it and regularly recycle all our Dapr pods otherwise it'll eventually run out of memory.

**Comment by dmitsh** [comment]
@slacksach, we've been working on improving RabbitMQ, and there was some progress. We missed the cutoff time to submit these changes in 1.5.0 release, but they will be available in the upcoming minor release 1.5.1.
In the meantime, I would suggest trying daprd from the master branch.

**Comment by slacksach** [comment]
thanks @dmitsh,

I wonder if this may also be caused when components are re-loaded due to a save or update, we see this in the logs, which is sometimes followed by connection growth but may be unrelated.

`{"app_id":"companies-api","instance":"companies-api-5df789d5f7-bqz58","level":"error","msg":"error from operator stream: rpc error: code = Unavailable desc = error reading from server: EOF","scope":"dapr.runtime","time":"2021-11-22T13:15:33.305689199Z","type":"log","ver":"1.5.0"}
`

**Comment by slacksach** [comment]
@dmitsh good news, have found the source (You may have already fixed this but logging here for others)

We are using https://fluxcd.io/docs/ as part of our GitOps process, every 10 minutes this runs against our GitHub repository then reconciles.

When it reconciles, it annotates all our managed objects, of which our RabbitMQ PubSub components are part,

This then triggers the Dapr Sidecar to reload all the components,
![image](https://user-images.githubusercontent.com/75119677/142909401-93e4617e-56e0-471b-9424-31a30f75d25b.png)

And hey presto, runaway dangling connection growth, the previous connections are not disposed and left dangling.

**Comment by yaron2** [comment]
When the items are reloaded, the operator will update the sidecar that a new component has been added. daprd does not check to see if this component is already loaded, and this explains the opening of new connections.

**Comment by slacksach** [comment]
For anyone else using Flux GitOps Toolkit v2, apply this annotation to your Dapr Component manifests which will prevent the issue until a fix is found.

```
  annotations:
    kustomize.toolkit.fluxcd.io/reconcile: disabled
```

**Comment by slacksach** [comment]
Can this be re-opened, it remains unsolved.

**Comment by dmitsh** [comment]
@slacksach , could you help us to find a minimalistic repro?

**Comment by paulyuk** [comment]
Hey @slacksach any chance we could get in touch with you?  I'd love to not only make sure you this issue settled, but I've also been hoping to learn more about Swoop's solution that builds on Dapr.  my microsoft.com email is paulyuk@, discord is paulyuk and my twitter is @paulyuki99.

**Comment by slacksach** [comment]
> Hey @slacksach any chance we could get in touch with you? I'd love to not only make sure you this issue settled, but I've also been hoping to learn more about Swoop's solution that builds on Dapr. my microsoft.com email is paulyuk@, discord is paulyuk and my twitter is @paulyuki99.

hi @paulyuk sure thing, I've dropped you an email, happy to discuss our solution and this issue.

**Comment by berndverst** [comment]
Should be fixed since 1.7 in PR https://github.com/dapr/dapr/pull/4213
